### PR TITLE
Add post of avoiding conditionals

### PR DESCRIPTION
**Pull request by tonysm**
### Added

- Blogpost about avoiding conditionals with some real world examples.

### CODE REVIEW

**Comment by tonysm** [comment]
[2017-08-08-avoiding-conditionals.pdf](https://github.com/madewithlove/blog/files/1206552/2017-08-08-avoiding-conditionals.pdf)

If you want to review it in PDF with embeded code, here you go.

File: _posts/2017-08-08-avoiding-conditionals.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+---
+layout: post
+title: "Avoiding Conditionals, an example"
+date: 2017-08-08
+authors: ["tony"]
+categories: [conditionals,collections,filter,map,loops]
+excerpt: Let's discuss an example I got from real code on how to avoid conditionals and end up having code that is actually easier to read.
+---
+
+I was not liking the code I wrote for a particular feature. It was "notify users when someone comments in a discussion", so ended up with the following requirements:
```
Comment (hannesvdvreken) [inline _posts/2017-08-08-avoiding-conditionals.md]: > I was not liking the code

I didn't like the code

(simple trick by [Carsten Wendt](http://carstensblog.com/))

Comment (hannesvdvreken) [inline _posts/2017-08-08-avoiding-conditionals.md]: > so ended up with the following requirements

just "it was xxx feature with these requirements:"?

The sentence seems off

File: _posts/2017-08-08-avoiding-conditionals.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    - the comment creator (because it doesn't make sense to notify them)
+    - the users that are assigned to the parent model associated with the discussion
+
+## Thinking in Conditionals
+
+I started implementing it the way I write the specs above. When I finished it, it felt a hacky and something I wasn't really happy with, see the pseudo code I ended up with:
+
+<script src="https://gist.github.com/tonysm/4f556246f9c9ca6b2eb60a2c6559963b.js"></script>
+
+There are just too many loops and conditionals and some duplication, really hacky, but I think it mirrors lots of code I've seen out there. Who haven't heard the phrase "just add an if statement here and you are good" (luckily for me its been a while - thanks madewithlove).
```
Comment (hannesvdvreken) [inline _posts/2017-08-08-avoiding-conditionals.md]: > its been a while

it has been a while (or short: it's been a while)

File: _posts/2017-08-08-avoiding-conditionals.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+## Thinking in Conditionals
+
+I started implementing it the way I write the specs above. When I finished it, it felt a hacky and something I wasn't really happy with, see the pseudo code I ended up with:
+
+<script src="https://gist.github.com/tonysm/4f556246f9c9ca6b2eb60a2c6559963b.js"></script>
+
+There are just too many loops and conditionals and some duplication, really hacky, but I think it mirrors lots of code I've seen out there. Who haven't heard the phrase "just add an if statement here and you are good" (luckily for me its been a while - thanks madewithlove). 
+
+I tried to _change_ this code, breaking it to small sub-methods. but then I kept passing the `$notifiedUsers` array as a reference just to avoid recreating a duplicate notification for the same users. That's when I decided to take the feature as another approach, lately I've been studying Functional Programming (FP), and inspired by Adam Wathan's ["Refactoring to Collections"](https://adamwathan.me/refactoring-to-collections/) book/videos, I decided to _refactor_ it using that approach.
```
Comment (hannesvdvreken) [inline _posts/2017-08-08-avoiding-conditionals.md]: > That's when I decided to take the feature as another approach

Maybe try:
That's when I decided to take a different approach to this feature

Comment (hannesvdvreken) [inline _posts/2017-08-08-avoiding-conditionals.md]: > . but then

. But then
(capital B)

Comment (hannesvdvreken) [inline _posts/2017-08-08-avoiding-conditionals.md]: > breaking it to small

breaking it into small

File: _posts/2017-08-08-avoiding-conditionals.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+Ok, that did the trick. Next, we have to make sure the comment creator doesn't get a notification for their own comments. Now that we have an array of notifications without duplication, I got it working like this:
+
+<script src="https://gist.github.com/tonysm/5c8697ab27b3313e246f411b556c6e72.js"></script>
+
+Great. Now, let's assemble everything together:
+
+<script src="https://gist.github.com/tonysm/271c2e598f2adf0b0de0aceaf38fbb7b.js"></script>
+
+Something I really liked and the code was really cleaner, it felt just write to me. I was happy with this. I think I'll be even happier if the we get the [pipe operator](https://wiki.php.net/rfc/pipe-operator), this could become something like this:
```
Comment (hannesvdvreken) [inline _posts/2017-08-08-avoiding-conditionals.md]: > it felt just write to me

it felt just right to me

File: _posts/2017-08-08-avoiding-conditionals.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+<script src="https://gist.github.com/tonysm/a8d34c896f045a2cbaf4d1e6a4b75be5.js"></script>
```
Comment (hannesvdvreken) [inline _posts/2017-08-08-avoiding-conditionals.md]: in this example:
> reduceDuplicates
what do you think about `removeDuplicates`?
Reducing, I think, is something else.

File: _posts/2017-08-08-avoiding-conditionals.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+## Conclusion
+
+The only downside I could think of would be if we have lots and lots of notifications or things to look after to generate the notifications. It's not a problem yet, because the model in question here is _time based_, so it has a defined existence timeline. A discussion will be created, debated and closed. When we end up having millions of users discussing in a single discussion to the point that it's consuming too much memory to build the notifications, I will be more than happy to _refactor_ this again.
+
+What do you think of this approach? Do you think it's better or worse than before? Do you know of different/better way of tackling this? Leave a comment bellow, I'm really interested in your opinions and experiences.
```
Comment (hannesvdvreken) [inline _posts/2017-08-08-avoiding-conditionals.md]: > Do you know of different/better way 

Do you know of different/better ways

Comment (hannesvdvreken) [inline _posts/2017-08-08-avoiding-conditionals.md]: > bellow

below

File: _posts/2017-08-08-avoiding-conditionals.md (hunk 5 above)
Comment (tonysm) [inline _posts/2017-08-08-avoiding-conditionals.md]: going to change it to `mergeDuplicates`

**Pull request merged**